StochasticBarrierFunctions/benchmarks/barrier_synthesis.jl

```diff
@@ -0,0 +1,286 @@
+    dim, σ =  config["dim"], config["σ"]
+    filename = "nonlinear/$(config["filename"])"
+    dataset = open_dataset(filename)
+    Xs = load_dynamics(dataset)
+    return dim, σ, Xs
+end
+
+function create_initial_region(config, dim)
+    # Initial region should never be empty
+    if haskey(config, "initial_region")
+        if haskey(config["initial_region"], "low") && haskey(config["initial_region"], "high")
+            return Hyperrectangle(low=config["initial_region"]["low"], high=config["initial_region"]["high"])
+        elseif haskey(config["initial_region"], "c") && haskey(config["initial_region"], "r")
+            center = config["initial_region"]["c"]
+            radius = config["initial_region"]["r"][1]
+            return Hyperrectangle(low=center .- radius, high=center .+ radius)
+        else
+            error("Invalid initial region configuration: expected 'low' and 'high' or 'c' and 'r'")
+        end
+    else
+        error("Initial region configuration is required and cannot be empty.")
+    end
+end
+
+function create_obstacle_region(config, dim)
+    if !haskey(config, "obstacle_region")
+        return EmptySet(dim)
+    end
+    
+    num_obstacles = config["obstacle_region"]["num_obstacles"]
+    obstacles = []
+    for i in 1:num_obstacles
+        obstacle_key = "obstacle_$i"
+        obstacle_data = config["obstacle_region"][obstacle_key]
+        
+        if haskey(obstacle_data, "c") && haskey(obstacle_data, "r")
+            # Construct a Hyperrectangle to represent Ball2
+            center = obstacle_data["c"]
+            radius = obstacle_data["r"][1]
+            push!(obstacles, Hyperrectangle(low = center .- radius, high = center .+ radius))
+        
+        elseif haskey(obstacle_data, "low") && haskey(obstacle_data, "high")
+            # Construct Hyperrectangle directly from low and high bounds
+            push!(obstacles, Hyperrectangle(low = obstacle_data["low"], high = obstacle_data["high"]))
+        
+        else
+            error("Invalid configuration for $obstacle_key: expected 'c' and 'r' or 'low' and 'high'")
+        end
+    end
+
+    return length(obstacles) > 1 ? UnionSet(obstacles...) : obstacles[1]
+end
+
+function generate_partitions(state_space, ϵ)
+    # Define ranges
+    ranges = [
+    range(
+        state_space.center[i] - state_space.radius[i],
+        step=ϵ[i],
+        length=Int(ceil((2 * state_space.radius[i]) / ϵ[i])) + 1
+    )
+    for i in 1:length(ϵ)
+    ]
+
+    # Generate a flat vector of Hyperrectangle objects for n-dimensions
+    state_partitions = [
+    Hyperrectangle(
+        low=[low for (low, high) in point_pairs],
+        high=[high for (low, high) in point_pairs]
+    )
+    for point_pairs in Iterators.product([zip(r[1:end-1], r[2:end]) for r in ranges]...)
+    ] |> vec
+
+    return state_partitions
+end
+
+function call_barrier_method(config, system_type_instance, barrier_type::SOS)
```

Comment:
That you dispatch this based on the `barrier_type` suggests to me you are using the wrong abstraction. I think it would be better if you have `system_factory` methods depending on the system type (could even do a Val-based dispatch) and `algorithm_factory` depending on the algorithm type (same with Val-based), both of which are called from a generic `call_barrier_method`. With this code structure, you end up with a single `call_barrier_method`, extensible factory methods (through Val-based dispatch), no `get_kwargs` and no `pwc_optimization_call`. 

I acknowledge that time is short before the deadline, so we could postpone this - but eventually the code really ought to be restructured this way.

Reply:
Good comment, I agree. Let's see if there is time before HSCC to do this.